ir/visitor.h

```diff
@@ -309,4 +309,53 @@ class P4WriteContext : public virtual Visitor {
     // as it might be an 'inout' access or it might be unable to decide.
 };
 
+
+/**
+ * Invoke an inspector @function for every node of type @NodeType in the subtree
+ * rooted at @root. The behavior is the same as a postorder Inspector.
+ */
+template <typename NodeType, typename Func>
+void forAllMatching(const IR::Node* root, Func&& function) {
+    struct NodeVisitor : public Inspector {
+        explicit NodeVisitor(Func&& function) : function(function) { }
+        Func function;
+        void postorder(const NodeType* node) override { function(node); }
+    };
+    root->apply(NodeVisitor(std::forward<Func>(function)));
+}
+
+/**
+ * Invoke a modifier @function for every node of type @NodeType in the subtree
+ * rooted at @root. The behavior is the same as a postorder Modifier.
+ *
+ * @return the root of the new, modified version of the subtree.
+ */
+template <typename NodeType, typename RootType, typename Func>
+const RootType* modifyAllMatching(const RootType* root, Func&& function) {
+    struct NodeVisitor : public Modifier {
+        explicit NodeVisitor(Func&& function) : function(function) { }
+        Func function;
+        bool postorder(NodeType* node) override { function(node); return true; }
```

Comment:
`Modifier::postorder` functions return void, not bool; I'm suprised this doesn't give a compile error?  Maybe its considered a covariant return type as bool > void?